### PR TITLE
Add 'string_view' unit tests.

### PR DESCRIPTION
**Pull request by spark008**

### CODE REVIEW

File: src/string_view_unittest.cc [hunk 1]
```diff
@@ -37,4 +37,177 @@ TEST(StringViewTest, Constructor_Data_Size) {
...
+
+  ASSERT_EQ(sv[0], 'h');
+}
+
+TEST(StringViewTest, RemovePrefixNone) {
+  const char cstring[] = "hello world";
+  string_view sv(cstring);
+
+  sv.remove_prefix(0);
+  ASSERT_STREQ(sv.data(), "hello world");
```
Comment (pwnall) [inline src/string_view_unittest.cc]: The test should be `ASSERT_EQ(sv.data(), data);`

The `string_view` name is unfortunate (much like the `string` name in `std::string`). This is effectively a handle to a pre-determined buffer, and remove_{suffix, prefix} are guaranteed to operate on that buffer.

File: src/string_view_unittest.cc [hunk 2]
```diff
@@ -37,4 +37,177 @@ TEST(StringViewTest, Constructor_Data_Size) {
...
+  ASSERT_EQ(sv.length(), 11);
+  ASSERT_EQ(sv.size(), 11);
+  ASSERT_EQ(sv.empty(), false);
+}
+
+TEST(StringViewTest, RemovePrefixPartial) {
```
Comment (pwnall) [inline src/string_view_unittest.cc]: 1 is an excellent boundary case test. Can you also add a test for a "mundane" value like 5?

File: src/string_view_unittest.cc [hunk 3]
```diff
@@ -37,4 +37,177 @@ TEST(StringViewTest, Constructor_Data_Size) {
...
+  ASSERT_EQ(sv1.compare(sv2), 0);
+  ASSERT_EQ(sv2.compare(sv1), 0);
+}
+
+TEST(StringViewTest, CompareLessThanLength) {
+  const char cstring[] = "hello world";
+  string_view sv1(cstring, 11);
+  string_view sv2(cstring, 15);
+
+  ASSERT_EQ(sv1.compare(sv2), -1);
```
Comment (pwnall) [inline src/string_view_unittest.cc]: This should be ASSERT_LT(..., 0). The spec doesn't promise `-1`.

http://en.cppreference.com/w/cpp/string/basic_string_view/compare

File: src/string_view_unittest.cc [hunk 4]
```diff
@@ -37,4 +37,177 @@ TEST(StringViewTest, Constructor_Data_Size) {
...
+  string_view sv2(cstring);
+
+  ASSERT_EQ(sv1.compare(sv2), 0);
+  ASSERT_EQ(sv2.compare(sv1), 0);
+}
+
+TEST(StringViewTest, CompareLessThanLength) {
+  const char cstring[] = "hello world";
+  string_view sv1(cstring, 11);
+  string_view sv2(cstring, 15);
```
Comment (pwnall) [inline src/string_view_unittest.cc]: This could technically crash, because you're creating a string_view covering memory you don't own.

How about making sv1 cover fewer than 11 chars instead here?

File: src/string_view_unittest.cc [hunk 5]
```diff
@@ -37,4 +37,177 @@ TEST(StringViewTest, Constructor_Data_Size) {
...
+
+  ASSERT_EQ(sv1.compare(sv2), -1);
+}
+
+TEST(StringViewTest, CompareGreaterThanLength) {
```
Comment (pwnall) [inline src/string_view_unittest.cc]: Seems like you could fold this case into the previous one using `ASSERT_GT(sv2.compare(sv1), 0);` -- same trick that you used in `CompareEqualTo`. The test name can be `CompareDifferentLengths`.

File: src/string_view_unittest.cc [hunk 6]
```diff
@@ -37,4 +37,177 @@ TEST(StringViewTest, Constructor_Data_Size) {
...
+TEST(StringViewTest, CompareLessThanValue) {
+  const char cstring1[] = "a";
+  string_view sv1(cstring1);
+  const char cstring2[] = "b";
+  string_view sv2(cstring2);
+
+  ASSERT_EQ(sv1.compare(sv2), -1);
+}
+
+TEST(StringViewTest, CompareGreaterThanValue) {
```
Comment (pwnall) [inline src/string_view_unittest.cc]: This is a good boundary case. Can you please also add a couple of cases comparing `aaaa` with `aaab` and `aaaa` with `aaba`?

**Comment by spark008** [comment]
@pwnall I hope I covered all of your comments. Let me know if I missed something.

File: src/string_view_unittest.cc [hunk 7]
```diff
@@ -37,4 +37,189 @@ TEST(StringViewTest, Constructor_Data_Size) {
...
+TEST(StringViewTest, CompareEqualTo) {
+  const char buffer[] = "hello world";
+  string_view sv1(buffer);
+  string_view sv2(buffer);
+
+  ASSERT_EQ(sv1.compare(sv2), 0);
+  ASSERT_EQ(sv2.compare(sv1), 0);
+}
+
+TEST(StringViewTest, CompareDifferentLengths) {
```
Comment (pwnall) [inline src/string_view_unittest.cc]: Can you please also add operator tests?

The easiest solution is that whenever you check something like `ASSERT_LT(sv1.compare(sv2), 0)` you also add a check like `ASSERT_LT(sv1, sv2)`

File: src/string_view_unittest.cc [hunk 8]
```diff
@@ -8,7 +8,7 @@
 
 namespace berrydb {
 
-TEST(StringViewTest, Constructor_Empty) {
+TEST(StringViewTest, ConstructorEmpty) {
   string_view sv;
 
   ASSERT_EQ(sv.data(), nullptr);
```
Comment (pwnall) [inline src/string_view_unittest.cc]: I messed up, because I've been used to Ruby. Actually, all of these ASSERTs should be EXPECTs. A find/replace should do the trick.

File: src/string_view_unittest.cc [hunk 9]
```diff
@@ -37,4 +37,189 @@ TEST(StringViewTest, Constructor_Data_Size) {
...
+  string_view sv2(buffer2);
+
+  ASSERT_LT(sv1.compare(sv2), 0);
+  ASSERT_GT(sv2.compare(sv1), 0);
+}
+
+TEST(StringViewTest, CompareWeightsValueOverLength) {
+  const char buffer1[] = "b";
+  string_view sv1(buffer1);
+  const char buffer2[] = {'a', '\0', '\0', '\0'};
```
Comment (pwnall) [inline src/string_view_unittest.cc]: Nice use of `\0`. You can also do `"a\0\0\0"` instead of an explicit array of chars if you think that's shorter.

**Comment by spark008** [comment]
@pwnall Hokey dokey I ended up adding entirely separate blocks for the inequality operators because I thought it was more systematic.

File: platform/berrydb/platform/string_view.h [hunk 10]
```diff
@@ -118,13 +118,13 @@ inline bool operator <(const string_view& l, const string_view& r) noexcept {
   return l.compare(r) < 0;
 }
 inline bool operator <=(const string_view& l, const string_view& r) noexcept {
-  return l.compare(r) < 0;
+  return l.compare(r) <= 0;
```
Comment (pwnall) [inline platform/berrydb/platform/string_view.h]: Oops. Thanks!

File: src/string_view_unittest.cc [hunk 11]
```diff
@@ -8,33 +8,295 @@
...
+
+  EXPECT_EQ(sv.data(), data);
+  EXPECT_EQ(sv.length(), 42);
+  EXPECT_EQ(sv.size(), 42);
+  EXPECT_EQ(sv.empty(), false);
+}
+
+TEST(StringViewTest, ElementAccessor) {
+  const char buffer[] = "hello world";
+  string_view sv(buffer);
```
Comment (pwnall) [inline src/string_view_unittest.cc]: Mosts tests should use the buffer/length constructor, not the C string constructor. Can you please add the length to the constructors where you use `buffer`?

For example: `string_view sv(buffer, 11);`
